### PR TITLE
use MSBuild binlog to report dependencies

### PR DESCRIPTION
**Pull request by brettfo**
### ~~This PR contains a temporary commit to redirect the smoke tests.  That commit will have to be removed prior to merging.~~ Temp commit has been removed.

This PR has a corresponding smoke test PR that will have to be merged at the same time: dependabot/smoke-tests#227

Previously, a temporary project was created to try and determine a project's full set of dependencies.  This wasn't always 100% accurate and could be slow due to all of the file copying.

This PR instead invokes MSBuild directly against the relevant project and produces a binary log (`.binlog`) that is then analyzed to get the full set of dependencies.

One of the benefits to this approach is that any give package can be directly associated with its parent project and MSBuild handles all of the complex property evaluation that might occur.

There are two main differences in behavior:

1. Properties are now fully evaluated; we no longer return `$(SomePackageVersion)`; we know exactly what was resolved.
2. We no longer report `Directory.Packages.props` as the source of a dependency.  While the dependency version might be found there, the dependency really lies with the project file.  This has no effect on actually performing the update; that occurs just like before.

Another relatively minor side effect is that we no longer report `NETStandard.Library` as a dependency; it's explicitly excluded.  While it _is_ a NuGet package that gets resolved, it's not one that can be directly updated so it is simply no longer reported.

The vast majority of the changes in this PR are to tests.

The primary file to review is `SdkProjectDiscovery.cs`; that's where the dependency analysis was completely migrated to the binary log.

### CODE REVIEW

File: nuget/helpers/lib/NuGetUpdater/.editorconfig [hunk 1]
```diff
@@ -20,6 +20,7 @@ tab_width = 4
 
 # New line preferences
 insert_final_newline = true
+#end_of_line = lf
```
Comment (brettfo) [inline nuget/helpers/lib/NuGetUpdater/.editorconfig:23]: I have my local environment set to always check out with `LF` so adding this makes it easy for me to `dotnet format`.

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Discover/SdkProjectDiscovery.cs [hunk 2]
```diff
@@ -1,129 +1,347 @@
...
+
+            // create dependencies
+            var tfms = packagesByTfm.Keys.OrderBy(tfm => tfm).ToImmutableArray();
+            var dependencies = tfms.SelectMany(tfm =>
+            {
+                return packagesByTfm[tfm].Keys.Select(packageName =>
+                {
+                    var packageVersion = packagesByTfm[tfm][packageName]!;
+                    var isTopLevel = topLevelPackageNames.Contains(packageName);
+                    var dependencyType = isTopLevel ? DependencyType.PackageReference : DependencyType.Unknown;
```
Comment (martincostello) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Discover/SdkProjectDiscovery.cs:187]: Will this deal with `GlobalPackageReference` items, or by the time MSBuild is processing it it'll have been turned into a normal `PackageReference`?

**Review by martincostello — COMMENTED** [review]
With MSBuild evaluation, will this improve (or fix) the way dependabot handles NuGet updates for projects that multi-target and/or prevent transient packages being updated unnecessarily?

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Discover/SdkProjectDiscovery.cs [hunk 3]
```diff
@@ -1,129 +1,347 @@
...
-        // Determine which targets and props files contribute to the build.
-        var (buildFiles, projectTargetFrameworks) = await MSBuildHelper.LoadBuildFilesAndTargetFrameworksAsync(repoRootPath, projectPath);
-        var tfms = projectTargetFrameworks.Order().ToImmutableArray();
+        // N.b., there are many paths used in this function.  The MSBuild binary log always reports fully qualified paths, so that's what will be used
+        // throughout until the very end when the appropriate kind of relative path is returned.
+
+        // step through the binlog one item at a time
+        var startingProjectDirectory = Path.GetDirectoryName(startingProjectPath)!;
+
+        // the following collectsion feature heavily; the shape is described as follows
```
Comment (sebasgomez238) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Discover/SdkProjectDiscovery.cs:44]: Nit: `collection`

**Review by sebasgomez238 — APPROVED** [review]

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Discover/SdkProjectDiscovery.cs (hunk 2 above)
Comment (brettfo) [reply nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Discover/SdkProjectDiscovery.cs:187]: `GlobalPackageReference` is turned into a `PackageReference` so we get this for free, but I've added a test with explicitly this scenario; good catch!

**Comment by brettfo** [comment]
> With MSBuild evaluation, will this improve (or fix) the way dependabot handles NuGet updates for projects that multi-target and/or prevent transient packages being updated unnecessarily?

The handling is certainly better.  When detecting dependencies we can now handle all `Condition` attributes no matter where they occur, but this PR is purely focused on the detection; there's still the issue of us being asked to update some package and making the change in the non-ideal or even wrong spot.  I'm discussing some features with the MSBuild team that could potentially allow more accurate location information for when we try to do the actual update, but that's future work.

**Comment by brettfo** [comment]
I've added tests explicitly checking:

1. Dependencies from repos using Central Package Versioning can be found.  I don't think we do the proper update yet (that'll be a follow-up PR), but we detect the proper package _and_ the relevant `Packages.props`.
2. Central Package Management when the `$(DirectoryPackagesPropsPath)` variable is overridden.  Same as above, we detect the package versions and the imported files properly, but we may not yet be able to do the update, but that's a future PR.

**Review by randhircs — APPROVED** [review]

**Pull request merged**